src/gui/log/logmodel.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2019  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QAbstractListModel>
+#include <QList>
+#include <QPair>
+#include <QString>
+
+#include "base/logger.h"
+
+class LogModel : public QAbstractListModel
+{
+    Q_OBJECT
+
+public:
+    LogModel(QList<QPair<QString, Log::MsgType>> &logLines, QObject *parent = nullptr);
```

Comment:
I would use struct instead of pair.

Reply:
The new struct would need to be seen by 3-4 different classes.
I prefer QPair, since it is easier to use in this case. I don't need to investigate in which header to declare the new struct to be seen by all other classes.

Reply:
> I don't need to investigate in which header to declare the new struct to be seen by all other classes.

You can produce its own header.
>The new struct would need to be seen by 3-4 different classes.

That's why it's bad! Using such abstract things as a pair in interfaces will lead to poorly readable (and maintainable) code. Isn't it? Pair is good in some general algorithms (algorithm templates).

Although, you can just brush me off here...

Reply:
Essentially it is **LogMessage** class/struct (which consists of the message itself and its type).
You can also use it in some other places, e.g. in `void LogModel::appendLine(const QString &line, const Log::MsgType &type);`

Reply:
I introduced a struct.